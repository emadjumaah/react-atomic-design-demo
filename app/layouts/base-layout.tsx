import React, { FunctionComponent } from 'react';
import { PageTitle } from 'elements';
import { Column, FlexSpace } from 'layouts';

export interface IBaseLayoutProps {
  /**
   * [required] Children should be an object that configures which components are injected into each
   * named slot
   */
  children: {
    pageTitle: string;
    content: JSX.Element;
  };
}

export const BaseLayout: FunctionComponent<IBaseLayoutProps> = props => (
  <Column padding={FlexSpace.XLARGE}>
    <PageTitle>{props.children.pageTitle}</PageTitle>
    {props.children.content}
  </Column>
);
