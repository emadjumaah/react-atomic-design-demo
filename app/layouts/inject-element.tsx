import React, { FunctionComponent } from 'react';
import { IDefaultProps } from 'types';
import { getChildKey } from 'layouts';

interface IInjectElementProps extends IDefaultProps {
  /**
   * [required] Element to inject between each child passed to component
   */
  inject: JSX.Element;
  /**
   * [required] Chilrend must be an array of JSX Elements
   */
  children: JSX.Element[];
}

export const InjectElement: FunctionComponent<IInjectElementProps> = props => (
  <>
    {React.Children.toArray(props.children).map(
      (child: JSX.Element, index: number) =>
        index < props.children.length - 1 ? (
          <React.Fragment key={getChildKey(child)}>
            {child}
            {props.inject}
          </React.Fragment>
        ) : (
          child
        ),
    )}
  </>
);
