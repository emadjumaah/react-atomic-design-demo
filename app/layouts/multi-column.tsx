import React, { FunctionComponent } from 'react';
import { IDefaultProps } from 'types';
import { Row, Column, getChildKey } from 'layouts';

interface IMultiColumnProps extends IDefaultProps {
  /**
   * [required] Children should be an array of elements that should exist within evenly
   * sized columns
   */
  children: JSX.Element[] | string[];
}

export const MultiColumn: FunctionComponent<IMultiColumnProps> = props => {
  return (
    <Row>
      {React.Children.toArray(props.children).map((child: JSX.Element) => (
        <Column data-testid="column-wrapper" key={getChildKey(child)}>
          {child}
        </Column>
      ))}
    </Row>
  );
};
