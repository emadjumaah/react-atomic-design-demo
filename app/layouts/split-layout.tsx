import React, { FunctionComponent } from 'react';
import { IDefaultProps } from 'types';
import { Row, Column } from 'layouts';

export enum SplitWidth {
  DEFAULT = '0 0 300px',
}

interface ISplitLayoutProps extends IDefaultProps {
  /**
   * [required] Children should be an object that configures which components are injected into each
   * named slot
   */
  children: {
    leftPanel: JSX.Element;
    rightPanel: JSX.Element;
  };

  /**
   * [optional] Controls width of left panel of SplitLayout
   *
   * @default SplitWidth.DEFAULT
   */
  leftPanelWidth?: SplitWidth;
}

export const SplitLayout: FunctionComponent<ISplitLayoutProps> = props => {
  const { leftPanel, rightPanel } = props.children;

  return (
    <Row>
      <Column flex={props.leftPanelWidth || SplitWidth.DEFAULT}>{leftPanel}</Column>
      <Column flex="1 1 auto">{rightPanel}</Column>
    </Row>
  );
};
