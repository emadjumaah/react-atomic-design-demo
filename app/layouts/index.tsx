import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import { IDefaultProps } from 'types';
import { InjectElement } from './inject-element';
import { isString } from 'utils';

export * from './base-layout';
export * from './split-layout';
export * from './inject-element';
export * from './multi-column';

const FlexElement = styled.div`
  display: flex;
`;

export const Hr = styled.hr`
  border-top: 1px solid var(--black);
  border-style: solid;
`;

export enum FlexPosition {
  START = 'flex-start',
  END = 'flex-end',
  CENTER = 'center',
}

export enum FlexSpace {
  SMALL = '8px',
  MEDIUM = '16px',
  LARGE = '32px',
  XLARGE = '64px',
  HOTDOG = '0px 16px',
  HAMBURGER = '16px 0px',
}

export enum DropShadow {
  DEFAULT = '3px 3px 3px var(--black)',
  NONE = 'none',
}

export enum Border {
  NONE = 'none',
  PRIMARY_LEFT = 'box-shadow: -5px 0px 0px var(--primary)',
  SECONDARY_LEFT = 'box-shadow: -5px 0px 0px var(--secondary)',
  FULL = 'border: 1px solid var(--black)',
  BOTTOM = 'border-bottom: 1px solid var(--black)',
}

export enum BorderRadius {
  NONE = '0px',
  BASE = '5px',
  CIRCLE = '100%',
}

interface IFlexElementProps extends IDefaultProps {
  /**
   * [optional] Padding to be applied to flex element
   */
  padding?: FlexSpace;
  /**
   * [optional] How content should be justified within flex element
   */
  justifyContent?: FlexPosition;
  /**
   * [optional] How content should be justified within flex element
   */
  alignItems?: FlexPosition;
  /**
   * [optional] Specifies width of flex element
   */
  width?: string;
  /**
   * [optional] Specifies height of flex element
   */
  height?: string;
  /**
   * [optional] Specifies spacing to be applied between children.  Children will be spaced
   * horizontally for Rows and vertically for Columns
   */
  childSpacing?: FlexSpace;
  /**
   * [optional] Specifies flex behavior.  Expects the same values as the `flex` css property:
   *
   *    flex: {grow} {shrink} {basis};
   */
  flex?: string;
  /**
   * [optional] Specifies dropshadow to show on element
   *
   * @default DropShadow.NONE
   */
  dropshadow?: DropShadow;
  /**
   * [optional] Enables hover effects for element
   *
   * Hover Effects include:
   *    cursor: pointer;
   */
  enableHoverEffects?: boolean;
  /**
   * [optional] Specifies border style to apply
   *
   * @default Border.NONE
   */
  border?: Border;
  /**
   * [optional] Alternative border to show on hover
   */
  borderOnHover?: Border;
  /**
   * [optional] Specifies border radius to apply
   *
   * @default Border.NONE
   */
  borderRadius?: BorderRadius;
}

const getHoverEffects = (borderOnHover: Border) => `
  &:hover {
    cursor: pointer;

    ${borderOnHover};
  }
`;

const getFlexStyles = (props: IFlexElementProps) => `
  ${!!props.padding ? `padding: ${props.padding};` : ''}
  ${!!props.justifyContent ? `justify-content: ${props.justifyContent};` : ''}
  ${!!props.alignItems ? `align-items: ${props.alignItems};` : ''}
  ${!!props.width ? `width: ${props.width};` : ''};
  ${!!props.height ? `height: ${props.height};` : ''};
  ${!!props.flex ? `flex: ${props.flex};` : ''};
`;

const getDesignStyles = (props: IFlexElementProps) => `
  ${!!props.dropshadow ? `box-shadow: ${props.dropshadow};` : ''};
  ${!!props.border ? props.border : ''};
  ${props.enableHoverEffects ? getHoverEffects(props.borderOnHover) : ''};
  border-radius: ${!!props.borderRadius ? props.borderRadius : BorderRadius.NONE};
`;

// #region Row Definition
const BaseRow: FunctionComponent<IFlexElementProps> = props => {
  if (!!props.childSpacing) {
    return (
      <FlexElement {...props}>
        <InjectElement inject={<Spacer width={props.childSpacing} />}>
          {props.children}
        </InjectElement>
      </FlexElement>
    );
  }
  return <FlexElement {...props} />;
};

export const Row = styled(BaseRow)<IFlexElementProps>`
  flex-direction: row;
  ${props => getFlexStyles(props)}
  ${props => getDesignStyles(props)}
`;
// #endregion Row Definition
// #region Column Definition
const BaseColumn: FunctionComponent<IFlexElementProps> = props => {
  if (!!props.childSpacing) {
    return (
      <FlexElement {...props}>
        <InjectElement inject={<Spacer height={props.childSpacing} />}>
          {props.children}
        </InjectElement>
      </FlexElement>
    );
  }
  return <FlexElement {...props} />;
};

export const Column = styled(BaseColumn)<IFlexElementProps>`
  flex-direction: column;
  ${props => getFlexStyles(props)}
  ${props => getDesignStyles(props)}
`;
// #endregion Column Definition

export interface ISpacerProps extends IDefaultProps {
  /**
   * [optional] Width spacer should take up
   */
  width?: string;
  /**
   * [optional] Height spacer should take up
   */
  height?: string;
  /**
   * [optional] Margin spacer should take up
   */
  margin?: string;
}

const getSpacerStyle = (props: ISpacerProps) => `
  ${!!props.width ? `width: ${props.width};` : ''}
  ${!!props.height ? `height: ${props.height};` : ''}
  ${!!props.margin ? `margin: ${props.margin};` : ''}
`;

export const Spacer = styled.div<ISpacerProps>`
  ${props => getSpacerStyle(props)}
`;

export const getChildKey = (child: JSX.Element | string) =>
  isString(child) ? child : child.key;
