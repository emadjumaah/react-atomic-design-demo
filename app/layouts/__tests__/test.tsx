import React from 'react';
import { render } from '@testing-library/react';
import { SplitLayout, MultiColumn, BaseLayout } from 'layouts';

describe('Layouts', () => {
  it('BaseLayout Renders', () => {
    const { container } = render(
      <BaseLayout>
        {{ pageTitle: 'Title Text', content: <div>Layout Content</div> }}
      </BaseLayout>,
    );
    expect(container).toMatchSnapshot();
  });
  it('SplitLayout Renders', () => {
    const { container } = render(
      <SplitLayout>
        {{ leftPanel: <div>Left Panel</div>, rightPanel: <div>Right Panel</div> }}
      </SplitLayout>,
    );
    expect(container).toMatchSnapshot();
  });
  it('MultiColumn Renders', () => {
    const { container, getAllByTestId } = render(
      <MultiColumn>{['First Column', 'Second Column', 'Third Column']}</MultiColumn>,
    );
    expect(container).toMatchSnapshot();
    expect(getAllByTestId('column-wrapper').length).toBe(3);
  });
});
