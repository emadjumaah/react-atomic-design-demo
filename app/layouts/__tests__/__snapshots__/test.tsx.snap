// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layouts BaseLayout Renders 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 64px;
  border-radius: 0px;
}

<div>
  <div
    class="c0 c1"
  >
    <h1>
      Title Text
    </h1>
    <div>
      Layout Content
    </div>
  </div>
</div>
`;

exports[`Layouts MultiColumn Renders 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  border-radius: 0px;
}

.c2 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-radius: 0px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2 c1"
      data-testid="column-wrapper"
    >
      First Column
    </div>
    <div
      class="c2 c1"
      data-testid="column-wrapper"
    >
      Second Column
    </div>
    <div
      class="c2 c1"
      data-testid="column-wrapper"
    >
      Third Column
    </div>
  </div>
</div>
`;

exports[`Layouts SplitLayout Renders 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  border-radius: 0px;
}

.c2 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 300px;
  -ms-flex: 0 0 300px;
  flex: 0 0 300px;
  border-radius: 0px;
}

.c3 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  border-radius: 0px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2 c1"
    >
      <div>
        Left Panel
      </div>
    </div>
    <div
      class="c3 c1"
    >
      <div>
        Right Panel
      </div>
    </div>
  </div>
</div>
`;
